### DISCUSSION TITLE
Quickstrom 0.5.0 Migration Thread

### DISCUSSION THREAD

**Discussion by owickstrom** [quickstrom/quickstrom#117] [General]
Hey everyone!

The new version is finally out and it's *quite different* from the
PureScript-based design I originally came up with. Needless to say,
lots of improvements in this new version, but also some things that
are missing or are fundamentally different from before.

I'm going to try capturing the fundamental changes below, but I also
suspect I'm going to miss out on certain details and use cases you
might have. So, let's do this as a thread and discuss whatever
problems come up in migrating from 0.4.x to 0.5.0.

Some big things:

* The language, called Specstrom, is new. It's something we (@liamoc
  and I) designed specifically for Quickstrom, with one foot in linear
  temporal logic and the other in web browsers and Javascript. 
  
  It's a two-level language, with QuickLTL formulae at the top and a
  dynamically typed functional language at the expression level.
  
  - Queries are expressed using special string literals (in backticks)
  instead of the `queryOne` and `queryAll` functions we had
  before.
  
  ```javascript
  let ~things = `.some-class`;
  ```

  - Attributes, properties, CSS styles and so on, are inferred from
    how you use elements in the specs. You don't have to list all
    the parts you want together with your query.

  ```javascript
  // .textContent will be read from all matching elements
  let things = `.some-class`;
  for thing in things { thing.textContent }
  ```
  
  - Specstrom has `null` (instead of PureScript's `Maybe`). It's also
    a special case in some functions, e.g. `for` and `map`, which
    implicitly filter out any `null` values:
    
    ```javascript
    for x in [1, 2, 3, 4] { x when x < 3 } == [1, 2]
    ```
    
  - Many useful functions are still missing. In the PureScript-based
    implementation we had to define functions for FFI calls. It was
    a lot of work, but we also got a lot of PureScript library code
    for free. Now, we have to build it all ourselves. PRs to the
    standard libraries of Specstrom and Quickstrom are welcome!
    
  - Documentation for the standard library is almost non-existent.
    @liamoc worked on a documentation generator so that we could
    have HTML API documentation. I'll look at integrating that with
    the Sphinx build for docs.quickstrom.io, and also just adding
    doc comments to the existing functions in the standard library.

* Actions cannot be sequenced as before. Instead you'll have to use
  preconditions to guard subsequent actions from being taken until
  a previous one has been taken:
  
  ```javascript
  action ~first! = ...; 
  action ~second! = ... when firstActionHasBeenTaken; // whatever that means
  ```
  
* The language is strict by default, laziness annotations are required
  if you don't want to eagerly evaluate something in the current
  state.
  
* The `next` operator no longer returns some value from the next
  state. If you want to have access to values from different states,
  you need to strictly bind one using `let`, and then refer to it
  from a formula inside a temporal operator:
  
  ```javascript
  let oldValue = value; // strictly evaluated in current state
  next (value > oldValue)
  ```
* Temporal operators have optional subscripts. These denote how many states they require to be read. The default is 100.

  ```javascript
  (always {50} ...) && next (... until {10} ...)
  ```
  
* There are three next-operators:

  - `next`: requires one more state to be ready, and evaluates its
    argument in that state
  - `nextT`: requires its argument in the next state if available,
    otherwise defaults to true
  - `nextF`: requires its argument in the next state if available,
    otherwise defaults to false
    
  If you nest temporal operators, you normally don't want to use
  `next`, because you can get into infinite loop situations.

These are some points I could come up with now. I'll add more as we
find them.

Let me know what you think!

**Comment by LiamGoodacre** [comment]
Is your `for x in [1, 2, 3, 4] { x when x < 3 } == [3, 4]` example not supposed to give `[1, 2]`, or am I reading it incorrectly?

**Reply by owickstrom** [reply]
Hah, nice catch. Thanks!
